File: core/protocol/go/send.go
package protocol

import (
	"bytes"
	"encoding/json"
	"strconv"
)

const (
	MaxFrameBytes    = 66048
	MaxSendBytes     = 61440
	MaxResultBytes   = 4096
	MaxMetadataBytes = 512
)

// SendCode adds frame errors without changing persisted errors. 新帧错误不改变旧错误契约。
type SendCode string

const (
	UnsupportedFrame    SendCode = "PROTOCOL_UNSUPPORTED_FRAME"
	CorrelationMismatch SendCode = "SEND_CORRELATION_MISMATCH"
	ResultConflict      SendCode = "SEND_RESULT_CONFLICT"
)

func (e SendCode) Error() string { return string(e) }

// Send carries caller intent, never sender authority. 发送意图不接受客户端身份声明。
type Send struct {
	ProtocolVersion uint32
	ClientMsgID     string
	ConversationID  string
	Version         uint32
	Type            string
	Payload         json.RawMessage
}

// Ack is a result shape; producers must already have committed. 仅已提交结果可产生此确认。
type Ack struct {
	ClientMsgID     string `json:"clientMsgId"`
	ConversationID  string `json:"conversationId"`
	SenderID        string `json:"senderId"`
	ServerMsgID     string `json:"serverMsgId"`
	ConversationSeq string `json:"conversationSeq"`
	ServerTime      string `json:"serverTime"`
}

// SendError carries a correlated stable code without diagnostics. 关联错误不携带敏感诊断。
type SendError struct {
	ClientMsgID    string `json:"clientMsgId"`
	ConversationID string `json:"conversationId"`
	Code           string `json:"code"`
}

// ServerFrame requires exactly one variant. 编码时要求恰好一个结果分支。
type ServerFrame struct {
	Ack     *Ack
	Error   *SendError
	Message *Message
}

// Supported indicates understood payload semantics, never permission. 语义支持并非发送授权。
func (s Send) Supported() bool { return s.ProtocolVersion == 1 && s.Type == "text" && s.Version == 1 }

type frameParts struct {
	version uint32
	kind    string
	raw     json.RawMessage
	fields  map[string]json.RawMessage
}

func objectFields(raw []byte) (map[string]json.RawMessage, error) {
	var f map[string]json.RawMessage
	if json.Unmarshal(raw, &f) != nil || f == nil {
		return nil, InvalidMessage
	}
	return f, nil
}
func stringField(f map[string]json.RawMessage, k string) (string, error) {
	var s string
	r, ok := f[k]
	if !ok || bytes.Equal(r, []byte("null")) || json.Unmarshal(r, &s) != nil {
		return "", InvalidMessage
	}
	return s, nil
}
func versionField(f map[string]json.RawMessage, k string) (uint32, error) {
	r := string(f[k])
	if !decimal(r, "2147483647") || r == "0" {
		return 0, InvalidMessage
	}
	n, _ := strconv.ParseUint(r, 10, 32)
	return uint32(n), nil
}
func forbidden(f map[string]json.RawMessage) bool {
	for _, k := range []string{"senderId", "serverMsgId", "conversationSeq", "serverTime", "status"} {
		if _, ok := f[k]; ok {
			return true
		}
	}
	return false
}
func parseFrame(wire []byte) (frameParts, error) {
	var p frameParts
	if err := strictJSONLimits(wire, MaxFrameBytes, MaxDepth+1); err != nil {
		return p, err
	}
	f, err := objectFields(wire)
	if err != nil {
		return p, err
	}
	if p.version, err = versionField(f, "protocolVersion"); err != nil {
		return p, err
	}
	if p.kind, err = stringField(f, "kind"); err != nil {
		return p, err
	}
	p.raw = f["body"]
	if p.fields, err = objectFields(p.raw); err != nil {
		return p, err
	}
	if p.kind == "send" && (forbidden(f) || forbidden(p.fields)) {
		return p, InvalidMessage
	}
	limit := MaxBytes
	if p.kind == "send" {
		limit = MaxSendBytes
	}
	if p.kind == "send_ack" || p.kind == "send_error" {
		limit = MaxResultBytes
	}
	if len(wire)-len(p.raw) > MaxMetadataBytes || len(p.raw) > limit {
		return p, TooLarge
	}
	if err = strictJSONLimits(p.raw, limit, MaxDepth); err != nil {
		return p, err
	}
	return p, nil
}
func sendShape(p frameParts) (Send, error) {
	s := Send{ProtocolVersion: p.version}
	var err error
	if s.ClientMsgID, err = stringField(p.fields, "clientMsgId"); err != nil {
		return s, err
	}
	if s.ConversationID, err = stringField(p.fields, "conversationId"); err != nil {
		return s, err
	}
	if s.Version, err = versionField(p.fields, "version"); err != nil {
		return s, err
	}
	if s.Type, err = stringField(p.fields, "type"); err != nil {
		return s, err
	}
	if !identifier(s.ClientMsgID, 128, false) || !identifier(s.ConversationID, 128, false) || !identifier(s.Type, 64, true) {
		return s, InvalidMessage
	}
	raw := p.fields["payload"]
	if _, err = objectFields(raw); err != nil {
		return s, err
	}
	s.Payload = append(json.RawMessage(nil), raw...)
	return s, nil
}
func ackShape(f map[string]json.RawMessage) (Ack, error) {
	var a Ack
	for k, d := range map[string]*string{"clientMsgId": &a.ClientMsgID, "conversationId": &a.ConversationID, "senderId": &a.SenderID, "serverMsgId": &a.ServerMsgID, "conversationSeq": &a.ConversationSeq, "serverTime": &a.ServerTime} {
		v, e := stringField(f, k)
		if e != nil {
			return a, e
		}
		*d = v
	}
	status, e := stringField(f, "status")
	if e != nil || status != "SERVER_PERSISTED" {
		return a, InvalidMessage
	}
	if !identifier(a.ClientMsgID, 128, false) || !identifier(a.ConversationID, 128, false) || !identifier(a.SenderID, 128, false) || !identifier(a.ServerMsgID, 128, false) || !decimal(a.ConversationSeq, "9223372036854775807") || !decimal(a.ServerTime, "9223372036854775807") {
		return a, InvalidMessage
	}
	return a, nil
}
func validCode(s string) bool {
	if len(s) == 0 || len(s) > 64 || s[0] < 'A' || s[0] > 'Z' {
		return false
	}
	for _, c := range []byte(s) {
		if !(c >= 'A' && c <= 'Z' || c >= '0' && c <= '9' || c == '_') {
			return false
		}
	}
	return true
}
func errorShape(f map[string]json.RawMessage) (SendError, error) {
	var e SendError
	var err error
	if e.ClientMsgID, err = stringField(f, "clientMsgId"); err != nil {
		return e, err
	}
	if e.ConversationID, err = stringField(f, "conversationId"); err != nil {
		return e, err
	}
	if e.Code, err = stringField(f, "code"); err != nil {
		return e, err
	}
	if !identifier(e.ClientMsgID, 128, false) || !identifier(e.ConversationID, 128, false) || !validCode(e.Code) {
		return e, InvalidMessage
	}
	return e, nil
}
func bodyShape(p frameParts) error {
	switch p.kind {
	case "send":
		_, e := sendShape(p)
		return e
	case "send_ack":
		_, e := ackShape(p.fields)
		return e
	case "send_error":
		_, e := errorShape(p.fields)
		return e
	case "message":
		_, e := decodeMessage(p.raw, false)
		return e
	}
	return nil
}
func sendSemantics(s Send) error {
	if s.Supported() {
		f, _ := objectFields(s.Payload)
		text, e := stringField(f, "text")
		if e != nil || len(text) == 0 || len(text) > MaxTextBytes {
			return InvalidMessage
		}
	}
	return nil
}

// DecodeSend accepts only client-direction frames. 只接受客户端发送方向。
func DecodeSend(wire []byte) (Send, error) {
	p, e := parseFrame(wire)
	if e != nil {
		return Send{}, e
	}
	if e = bodyShape(p); e != nil {
		return Send{}, e
	}
	if p.version != 1 {
		return Send{}, UnsupportedVersion
	}
	if p.kind != "send" {
		return Send{}, UnsupportedFrame
	}
	s, e := sendShape(p)
	if e == nil {
		e = sendSemantics(s)
	}
	if e != nil {
		return Send{}, e
	}
	return s, nil
}

// DecodeServerFrame requires explicit server framing. 服务端帧不猜测裸消息格式。
func DecodeServerFrame(wire []byte) (ServerFrame, error) {
	var f ServerFrame
	p, e := parseFrame(wire)
	if e != nil {
		return f, e
	}
	if e = bodyShape(p); e != nil {
		return f, e
	}
	if p.version != 1 {
		return f, UnsupportedVersion
	}
	switch p.kind {
	case "send_ack":
		a, e := ackShape(p.fields)
		f.Ack = &a
		return f, e
	case "send_error":
		v, e := errorShape(p.fields)
		f.Error = &v
		return f, e
	case "message":
		m, e := Decode(p.raw)
		if e != nil {
			return f, e
		}
		f.Message = &m
		return f, nil
	default:
		return f, UnsupportedFrame
	}
}
func lowerBound(limit int, sizes ...int) error {
	total := 0
	for _, n := range sizes {
		if n > limit-total {
			return TooLarge
		}
		total += n
	}
	return nil
}
func wrap(version uint32, kind string, body []byte) []byte {
	return []byte(`{"protocolVersion":` + strconv.FormatUint(uint64(version), 10) + `,"kind":"` + kind + `","body":` + string(body) + `}`)
}

// Match JSON wire escaping rather than HTML/script escaping, including invalid public
// identifiers, so size-versus-shape errors agree with the Rust encoder.
func quotedWireString(out []byte, value string) []byte {
	out = append(out, '"')
	const hex = "0123456789abcdef"
	for i := 0; i < len(value); i++ {
		ch := value[i]
		switch ch {
		case '"', '\\':
			out = append(out, '\\', ch)
		case '\b':
			out = append(out, '\\', 'b')
		case '\f':
			out = append(out, '\\', 'f')
		case '\n':
			out = append(out, '\\', 'n')
		case '\r':
			out = append(out, '\\', 'r')
		case '\t':
			out = append(out, '\\', 't')
		default:
			if ch < 32 {
				out = append(out, '\\', 'u', '0', '0', hex[ch>>4], hex[ch&15])
			} else {
				out = append(out, ch)
			}
		}
	}
	return append(out, '"')
}
func stringObject(pairs ...[2]string) []byte {
	out := []byte{'{'}
	for i, pair := range pairs {
		if i > 0 {
			out = append(out, ',')
		}
		out = quotedWireString(out, pair[0])
		out = append(out, ':')
		out = quotedWireString(out, pair[1])
	}
	return append(out, '}')
}

// EncodeSend validates public fields and keeps payload tokens. 公有字段编码仍须验证。
func EncodeSend(s Send) ([]byte, error) {
	if e := lowerBound(MaxSendBytes, len(s.ClientMsgID), len(s.ConversationID), len(s.Type), len(s.Payload)); e != nil {
		return nil, e
	}
	// Validate the independent fragment before concatenation; no injected frame fields.
	if s.Payload != nil && !json.Valid(s.Payload) {
		return nil, InvalidJSON
	}
	meta := stringObject([2]string{"clientMsgId", s.ClientMsgID}, [2]string{"conversationId", s.ConversationID}, [2]string{"type", s.Type})
	body := append(meta[:len(meta)-1], []byte(`,"version":`+strconv.FormatUint(uint64(s.Version), 10)+`,"payload":`)...)

	if s.Payload == nil {
		body = append(body, []byte("null")...)
	} else {
		body = append(body, s.Payload...)
	}
	body = append(body, '}')
	wire := wrap(s.ProtocolVersion, "send", body)
	if _, e := DecodeSend(wire); e != nil {
		return nil, e
	}
	return wire, nil
}

// EncodeServerFrame validates exactly one result. 输出只允许一个合法服务端结果。
func EncodeServerFrame(f ServerFrame) ([]byte, error) {
	n := 0
	if f.Ack != nil {
		n++
	}
	if f.Error != nil {
		n++
	}
	if f.Message != nil {
		n++
	}
	if n != 1 {
		return nil, InvalidMessage
	}
	var body []byte
	var e error
	kind := ""
	if a := f.Ack; a != nil {
		kind = "send_ack"
		if e = lowerBound(MaxResultBytes, len(a.ClientMsgID), len(a.ConversationID), len(a.SenderID), len(a.ServerMsgID), len(a.ConversationSeq), len(a.ServerTime)); e != nil {
			return nil, e
		}
		body = stringObject([2]string{"status", "SERVER_PERSISTED"}, [2]string{"clientMsgId", a.ClientMsgID}, [2]string{"conversationId", a.ConversationID}, [2]string{"senderId", a.SenderID}, [2]string{"serverMsgId", a.ServerMsgID}, [2]string{"conversationSeq", a.ConversationSeq}, [2]string{"serverTime", a.ServerTime})
	}
	if v := f.Error; v != nil {
		kind = "send_error"
		if e = lowerBound(MaxResultBytes, len(v.ClientMsgID), len(v.ConversationID), len(v.Code)); e != nil {
			return nil, e
		}
		body = stringObject([2]string{"clientMsgId", v.ClientMsgID}, [2]string{"conversationId", v.ConversationID}, [2]string{"code", v.Code})
	}
	if f.Message != nil {
		kind = "message"
		body, e = Encode(*f.Message)
	}
	if e != nil {
		return nil, e
	}
	wire := wrap(1, kind, body)
	if _, e = DecodeServerFrame(wire); e != nil {
		return nil, e
	}
	return wire, nil
}

// Disposition is a local conservative retry decision. 未知错误保守停止自动重试。
type RetryDisposition string

const (
	RetrySameIntent    RetryDisposition = "RetrySameIntent"
	AuthRecovery       RetryDisposition = "AuthRecovery"
	StopAutomaticRetry RetryDisposition = "StopAutomaticRetry"
)

func Disposition(code string) RetryDisposition {
	switch code {
	case "SERVER_TEMPORARY_UNAVAILABLE":
		return RetrySameIntent
	case "AUTH_REQUIRED", "AUTH_TOKEN_EXPIRED":
		return AuthRecovery
	default:
		return StopAutomaticRetry
	}
}

File: core/protocol/rust/src/send.rs
//! 发送与持久化回执帧；解析不证明提交 / Send and result frames; parsing does not prove commit.
use std::{collections::BTreeMap, fmt};

use serde_json::value::{RawValue, to_raw_value};

pub use crate::intent::{compare_results, correlate, same_intent};
use crate::{Message, Object, decimal, field, id, object, scan, string, version};

pub const MAX_FRAME_BYTES: usize = 66_048;
pub const MAX_SEND_BYTES: usize = 61_440;
pub const MAX_RESULT_BYTES: usize = 4_096;
pub const MAX_METADATA_BYTES: usize = 512;

/// 保留旧错误枚举兼容性 / Preserve the existing codec error enum.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Error {
    Protocol(crate::Error),
    UnsupportedFrame,
    CorrelationMismatch,
    ResultConflict,
}
impl Error {
    pub const fn code(self) -> &'static str {
        match self {
            Self::Protocol(error) => error.code(),
            Self::UnsupportedFrame => "PROTOCOL_UNSUPPORTED_FRAME",
            Self::CorrelationMismatch => "SEND_CORRELATION_MISMATCH",
            Self::ResultConflict => "SEND_RESULT_CONFLICT",
        }
    }
}
impl From<crate::Error> for Error {
    fn from(error: crate::Error) -> Self {
        Self::Protocol(error)
    }
}
impl fmt::Display for Error {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.code())
    }
}
impl std::error::Error for Error {}

/// 发送意图，不包含服务器权威字段 / Send intent without server-authoritative fields.
#[derive(Clone, Debug)]
pub struct Send {
    pub protocol_version: u32,
    pub client_msg_id: String,
    pub conversation_id: String,
    pub version: u32,
    pub message_type: String,
    pub payload: Box<RawValue>,
}

impl Send {
    /// 是否识别发送 schema（非授权）/ Whether the send schema is understood, not authorized.
    pub fn supported(&self) -> bool {
        self.protocol_version == 1
            && self.version == 1
            && matches!(self.message_type.as_str(), "text" | "media")
    }
}

/// 仅代表 SERVER_PERSISTED；生产者须保证事务已提交 / Producer must ensure transaction commit.
#[derive(Clone, Debug)]
pub struct Ack {
    pub client_msg_id: String,
    pub conversation_id: String,
    pub sender_id: String,
    pub server_msg_id: String,
    pub conversation_seq: String,
    pub server_time: String,
}
/// 仅可关联可信解析的请求；未知错误码保留 / Correlate only parsed requests; retain unknown codes.
#[derive(Clone, Debug)]
pub struct SendError {
    pub client_msg_id: String,
    pub conversation_id: String,
    pub code: String,
}
/// 明确区分持久结果与失败；错误不能撤销持久化 / Errors cannot revoke a persisted result.
#[derive(Clone, Debug)]
pub enum ServerFrame {
    Ack(Ack),
    Error(SendError),
    Message(Message),
}

/// 本地保守重试分类；调度和有限退避由 SDK 负责 / SDK owns scheduling and bounded backoff.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Disposition {
    RetrySameIntent,
    AuthRecovery,
    StopAutomaticRetry,
}
pub fn disposition(code: &str) -> Disposition {
    match code {
        "SERVER_TEMPORARY_UNAVAILABLE" => Disposition::RetrySameIntent,
        "AUTH_REQUIRED" | "AUTH_TOKEN_EXPIRED" => Disposition::AuthRecovery,
        _ => Disposition::StopAutomaticRetry,
    }
}

fn invalid<T>() -> Result<T, Error> {
    Err(crate::Error::InvalidMessage.into())
}
fn identifiers(values: &[&str]) -> Result<(), Error> {
    if values.iter().all(|value| id(value)) {
        Ok(())
    } else {
        invalid()
    }
}
fn forbidden(fields: &Object) -> bool {
    [
        "senderId",
        "serverMsgId",
        "conversationSeq",
        "serverTime",
        "status",
    ]
    .iter()
    .any(|key| fields.contains_key(*key))
}
fn code_valid(code: &str) -> bool {
    !code.is_empty()
        && code.len() <= 64
        && code.as_bytes()[0].is_ascii_uppercase()
        && code
            .bytes()
            .all(|c| c.is_ascii_uppercase() || c.is_ascii_digit() || c == b'_')
}
fn send_shape(mut fields: Object, protocol_version: u32) -> Result<Send, Error> {
    let send = Send {
        protocol_version,
        client_msg_id: string(&fields, "clientMsgId")?,
        conversation_id: string(&fields, "conversationId")?,
        version: version(&fields, "version")?,
        message_type: string(&fields, "type")?,
        payload: fields
            .remove("payload")
            .ok_or(crate::Error::InvalidMessage)?,
    };
    identifiers(&[&send.client_msg_id, &send.conversation_id])?;
    if !crate::message_type(&send.message_type) {
        return invalid();
    }
    object(send.payload.get())?;
    Ok(send)
}
fn send_semantics(send: &Send) -> Result<(), Error> {
    if send.version != 1 {
        return Ok(());
    }
    match send.message_type.as_str() {
        "text" => {
            let text = string(&object(send.payload.get())?, "text")?;
            if text.is_empty() || text.len() > crate::MAX_TEXT_BYTES {
                return invalid();
            }
        }
        "media" => crate::media::validate(send.payload.get())?,
        _ => {}
    }
    Ok(())
}
fn ack_shape(fields: &Object) -> Result<Ack, Error> {
    if string(fields, "status")? != "SERVER_PERSISTED" {
        return invalid();
    }
    let ack = Ack {
        client_msg_id: string(fields, "clientMsgId")?,
        conversation_id: string(fields, "conversationId")?,
        sender_id: string(fields, "senderId")?,
        server_msg_id: string(fields, "serverMsgId")?,
        conversation_seq: string(fields, "conversationSeq")?,
        server_time: string(fields, "serverTime")?,
    };
    identifiers(&[
        &ack.client_msg_id,
        &ack.conversation_id,
        &ack.sender_id,
        &ack.server_msg_id,
    ])?;
    if !decimal(&ack.conversation_seq) || !decimal(&ack.server_time) {
        return invalid();
    }
    Ok(ack)
}
fn error_shape(fields: &Object) -> Result<SendError, Error> {
    let error = SendError {
        client_msg_id: string(fields, "clientMsgId")?,
        conversation_id: string(fields, "conversationId")?,
        code: string(fields, "code")?,
    };
    identifiers(&[&error.client_msg_id, &error.conversation_id])?;
    if !code_valid(&error.code) {
        return invalid();
    }
    Ok(error)
}

enum Frame {
    Send(Send),
    Server(ServerFrame),
    Unknown,
}
fn decode_frame(input: &[u8], client_direction: bool) -> Result<Frame, Error> {
    scan::check_with_limits(input, MAX_FRAME_BYTES, crate::MAX_DEPTH + 1)?;
    let text = std::str::from_utf8(input).map_err(|_| crate::Error::InvalidJson)?;
    let outer = object(text)?;
    let outer_version = version(&outer, "protocolVersion")?;
    let kind = string(&outer, "kind")?;
    let raw_body = field(&outer, "body")?.get();
    let body = object(raw_body)?;
    if kind == "send" && (forbidden(&outer) || forbidden(&body)) {
        return invalid();
    }
    let body_budget = match kind.as_str() {
        "send" => MAX_SEND_BYTES,
        "send_ack" | "send_error" => MAX_RESULT_BYTES,
        _ => crate::MAX_WIRE_BYTES,
    };
    if input.len() - raw_body.len() > MAX_METADATA_BYTES || raw_body.len() > body_budget {
        return Err(crate::Error::MessageTooLarge.into());
    }
    scan::check_with_limits(raw_body.as_bytes(), body_budget, crate::MAX_DEPTH)?;
    let frame = match kind.as_str() {
        "send" => Frame::Send(send_shape(body, outer_version)?),
        "send_ack" => Frame::Server(ServerFrame::Ack(ack_shape(&body)?)),
        "send_error" => Frame::Server(ServerFrame::Error(error_shape(&body)?)),
        "message" => Frame::Server(ServerFrame::Message(crate::decode_shape(raw_body)?)),
        _ => Frame::Unknown,
    };
    if outer_version != 1 {
        return Err(crate::Error::UnsupportedVersion.into());
    }
    match &frame {
        Frame::Send(send) if client_direction => send_semantics(send)?,
        Frame::Server(ServerFrame::Message(message)) if !client_direction => {
            crate::validate_semantics(message)?
        }
        Frame::Server(_) if !client_direction => (),
        _ => return Err(Error::UnsupportedFrame),
    }
    Ok(frame)
}

/// 完整形状先于版本和方向校验 / Complete shape precedes version and direction checks.
pub fn decode_send(input: &[u8]) -> Result<Send, Error> {
    match decode_frame(input, true)? {
        Frame::Send(send) => Ok(send),
        _ => Err(Error::UnsupportedFrame),
    }
}
/// 仅解析服务器方向帧，不猜测裸消息 / Decode server frames without guessing bare messages.
pub fn decode_server_frame(input: &[u8]) -> Result<ServerFrame, Error> {
    match decode_frame(input, false)? {
        Frame::Server(frame) => Ok(frame),
        _ => Err(Error::UnsupportedFrame),
    }
}

type Output = BTreeMap<&'static str, Box<RawValue>>;
fn strings(fields: &[(&'static str, &str)], budget: usize) -> Result<Output, Error> {
    if fields
        .iter()
        .map(|(_, value)| value.len())
        .fold(0usize, usize::saturating_add)
        > budget
    {
        return Err(crate::Error::MessageTooLarge.into());
    }
    fields
        .iter()
        .map(|(name, value)| {
            Ok((
                *name,
                to_raw_value(value).map_err(|_| crate::Error::InvalidMessage)?,
            ))
        })
        .collect()
}
fn wrap(protocol_version: u32, kind: &str, body: Box<RawValue>) -> Result<Vec<u8>, Error> {
    let mut outer = Output::new();
    outer.insert(
        "protocolVersion",
        to_raw_value(&protocol_version).map_err(|_| crate::Error::InvalidMessage)?,
    );
    outer.insert(
        "kind",
        to_raw_value(kind).map_err(|_| crate::Error::InvalidMessage)?,
    );
    outer.insert("body", body);
    serde_json::to_vec(&outer).map_err(|_| crate::Error::InvalidJson.into())
}

/// 重验公共字段且保留原始 payload 数值 / Revalidate public fields without changing raw numbers.
pub fn encode_send(send: &Send) -> Result<Vec<u8>, Error> {
    let minimum = [
        send.client_msg_id.len(),
        send.conversation_id.len(),
        send.message_type.len(),
        send.payload.get().len(),
    ]
    .into_iter()
    .fold(0usize, usize::saturating_add);
    if minimum > MAX_SEND_BYTES {
        return Err(crate::Error::MessageTooLarge.into());
    }
    // RawValue guarantees an independent JSON fragment; full-frame scanning below
    // preserves size-before-duplicate/depth precedence even after string escaping.
    let mut body = strings(
        &[
            ("clientMsgId", &send.client_msg_id),
            ("conversationId", &send.conversation_id),
            ("type", &send.message_type),
        ],
        MAX_SEND_BYTES,
    )?;
    body.insert(
        "version",
        to_raw_value(&send.version).map_err(|_| crate::Error::InvalidMessage)?,
    );
    body.insert("payload", send.payload.clone());
    let encoded = wrap(
        send.protocol_version,
        "send",
        to_raw_value(&body).map_err(|_| crate::Error::InvalidMessage)?,
    )?;
    decode_send(&encoded)?;
    Ok(encoded)
}
/// 重验结果字段；编码成功不证明服务端提交 / Revalidate fields; encoding never proves commit.
pub fn encode_server_frame(frame: &ServerFrame) -> Result<Vec<u8>, Error> {
    let (kind, body) = match frame {
        ServerFrame::Ack(ack) => (
            "send_ack",
            strings(
                &[
                    ("status", "SERVER_PERSISTED"),
                    ("clientMsgId", &ack.client_msg_id),
                    ("conversationId", &ack.conversation_id),
                    ("senderId", &ack.sender_id),
                    ("serverMsgId", &ack.server_msg_id),
                    ("conversationSeq", &ack.conversation_seq),
                    ("serverTime", &ack.server_time),
                ],
                MAX_RESULT_BYTES,
            )?,
        ),
        ServerFrame::Error(error) => (
            "send_error",
            strings(
                &[
                    ("clientMsgId", &error.client_msg_id),
                    ("conversationId", &error.conversation_id),
                    ("code", &error.code),
                ],
                MAX_RESULT_BYTES,
            )?,
        ),
        ServerFrame::Message(message) => {
            let raw = String::from_utf8(crate::encode(message)?)
                .map_err(|_| crate::Error::InvalidJson)?;
            let encoded = wrap(
                1,
                "message",
                RawValue::from_string(raw).map_err(|_| crate::Error::InvalidJson)?,
            )?;
            decode_server_frame(&encoded)?;
            return Ok(encoded);
        }
    };
    let encoded = wrap(
        1,
        kind,
        to_raw_value(&body).map_err(|_| crate::Error::InvalidMessage)?,
    )?;
    decode_server_frame(&encoded)?;
    Ok(encoded)
}
